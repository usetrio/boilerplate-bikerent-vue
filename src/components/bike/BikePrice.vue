<script lang="ts">
import { defineComponent, type PropType } from 'vue'

import type { CurrencyCode } from '@/core/config'
import { formatCurrency } from '@/core/helpers/currency'

export type PropRateType = 'daily' | 'weekly' | 'monthly' | 'anually' | 'none'

export default defineComponent({
  name: 'BikePrice',
  props: {
    price: {
      type: Number,
      required: true
    },
    currency: {
      type: String as PropType<CurrencyCode>,
      required: true
    },
    rate: {
      type: String as PropType<PropRateType>,
      default: ''
    }
  },
  computed: {
    formattedPrice() {
      return formatCurrency(this.price, this.currency)
    },
    rateLabel() {
      let label = ''
      switch (this.rate) {
        case 'daily':
          label = 'Day'
          break
        case 'weekly':
          label = 'Week'
          break
        case 'monthly':
          label = 'Month'
          break
        case 'anually':
          label = 'Year'
          break
        default:
      }

      return label
    }
  }
})
</script>

<template>
  <div class="bike__price">
    <strong>{{ formattedPrice }}</strong>
    <template v-if="rateLabel">/ {{ rateLabel }}</template>
  </div>
</template>

<style lang="scss" scoped>
@include block('bike__price') {
  font-weight: 600;
  font-size: 16px;
  line-height: 23px;

  strong {
    font-weight: 800;
    font-size: 24px;
    line-height: 31px;
  }
}
</style>
