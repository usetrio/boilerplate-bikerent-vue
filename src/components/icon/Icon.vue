<script lang="ts">
import { defineComponent, type PropType } from 'vue'

export enum IconType {
  REGULAR = 'regular',
  SOLID = 'solid'
}

export type PropIconType = `${IconType}`

export default defineComponent({
  name: 'Icon',
  props: {
    type: {
      type: String as PropType<PropIconType>,
      default: IconType.REGULAR
    },
    size: {
      type: String as PropType<UISizing | UISizingType>,
      default: 'base'
    }
  },
  computed: {
    icon() {
      const [icon] = (this.$slots.default && this.$slots.default()) || []

      if (!icon) {
        return ''
      }

      return (icon.children as String).trim()
    },
    componentClass() {
      return ['icon', `text-${this.size}`]
    },
    computedClass() {
      const classes = [`fa-${this.type}`]

      if (this.icon) {
        classes.push(`fa-${this.icon}`)
      }

      return classes
    }
  }
})
</script>

<template>
  <div :class="componentClass">
    <em :class="computedClass" />
  </div>
</template>

<style lang="scss" scoped>
.icon {
  @apply .inline-block;
  position: relative;
}
</style>
