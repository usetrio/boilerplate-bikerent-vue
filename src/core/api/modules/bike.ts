import type { BikeRent, BikeRentDetails, BikeReturnDetails } from './typings/bike'

// import { ApiResponseError, client } from '@/core/api'
import { client } from '@/core/api'

export type { BikeRent, BikeRentDetails, BikeReturnDetails }

/**
 * List Bikes
 *
 * @param {boolean} [available=false] List only available bikes for renting
 *
 * @returns {Promise<Bike[]>}
 */
export async function list(available = false): Promise<Bike[]> {
  return available ? await listAvailable() : await listAll()
}

/**
 * Lists all bikes.
 *
 * @returns {Promise<Bike[]>}
 */
export async function listAll(): Promise<Bike[]> {
  const { data } = await client.get<Bike[]>('bikes')

  return data
}

/**
 * Lists only available bikes, bikes that are not rented.
 *
 * @returns {Promise<Bike[]>}
 */
export async function listAvailable(): Promise<Bike[]> {
  const { data } = await client.get<Bike[]>('bikes/available')

  return data
}

/**
 * Rents a bike; you can also book future rents.
 *
 *
 * @param details
 * @returns
 */
export async function rent(details: BikeRentDetails): Promise<BikeRent> {
  const { data } = await client.post<BikeRent>('bikes/rent', details)

  return data
}

/**
 * Returns a bike
 * @param {BikeReturnDetails} details
 */
export async function returns(details: BikeReturnDetails): Promise<void> {
  await client.post('bikes/return', details)
}
