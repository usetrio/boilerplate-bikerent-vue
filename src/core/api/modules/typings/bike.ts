export interface BikeRent {
  rentAmount: number
  fee: number
  totalAmount: number
}

export interface BikeRentDetails {
  bikeId: number
  userId: number
  dateFrom: string
  dateTo: string
}

export interface BikeReturnDetails {
  bikeId: number
  userId: number
}
