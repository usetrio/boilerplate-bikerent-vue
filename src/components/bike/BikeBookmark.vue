<script lang="ts">
import { Icon, IconType } from '@/components/icon'
import { defineComponent, type PropType } from 'vue'

export default defineComponent({
  name: 'BikeBookmark',
  components: {
    Icon
  },
  props: {
    active: {
      type: Boolean,
      default: false
    },
    outlined: {
      type: Boolean,
      default: false
    },
    size: {
      type: String as PropType<UISizing | UISizingType>,
      default: 'base'
    },
    width: {
      type: Number,
      default: 0
    }
  },
  emits: {
    'update:active': (_state: boolean) => true
  },
  computed: {
    iconType() {
      return this.active ? IconType.SOLID : IconType.REGULAR
    },
    componentClasses() {
      const classes: UICSSClass = ['bike-bookmark']

      classes.push({ 'bike-bookmark--outlined': this.outlined, 'bike-bookmark--active': this.active })

      return classes
    },
    componentStyle() {
      const style: UICSSProperties = {}

      if (this.width) {
        style.width = `${this.width}px`
      }

      return style
    }
  },
  methods: {
    handleToggleActive() {
      this.$emit('update:active', !this.active)
    }
  }
})
</script>

<template>
  <div :class="componentClasses" :style="componentStyle" @click.prevent="handleToggleActive">
    <icon :type="iconType" :size="size">heart</icon>
  </div>
</template>

<style lang="scss" scoped>
.bike-bookmark {
  position: relative;
  cursor: pointer;
  aspect-ratio: 1;

  @include modifier('outlined') {
    @apply .p-3, .inline-flex, .items-center, .justify-center;

    border: 1px solid get-theme-color('gray');
    border-radius: $border-radius-md;
  }
}
</style>
