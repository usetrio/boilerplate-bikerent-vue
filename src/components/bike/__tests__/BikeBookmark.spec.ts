import { describe, expect, it } from 'vitest'
import { flushPromises, mount } from '@vue/test-utils'

import { BikeBookmark } from '@/components/bike'
import { Icon } from '@/components/icon'

describe('BikeBookmark', () => {
  it('renders correctly', () => {
    const wrapper = mount(BikeBookmark)

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('shows active state', () => {
    const wrapper = mount(BikeBookmark, { props: { active: true } })

    expect(wrapper.classes()).contain('bike-bookmark--active')
  })

  it('show outlined state', async () => {
    const wrapper = mount(BikeBookmark, { props: { outlined: true } })

    expect(wrapper.classes()).contain('bike-bookmark--outlined')
  })

  it('shows different width', async () => {
    const wrapper = mount(BikeBookmark, { props: { width: 150 } })

    expect(wrapper.attributes('style')).contain('width: 150px')
  })

  it('shows different sizing', async () => {
    // TODO: fix vitest config to handle global typings
    // const wrapper = mount(BikeBookmark, { props: { size: UISizing.LG } })
    const wrapper = mount(BikeBookmark, { props: { size: 'lg' } })

    expect(wrapper.findComponent(Icon).classes()).contain('text-lg')
  })

  it('changes active state', async () => {
    const wrapper = mount(BikeBookmark, { props: { active: true } })

    expect(wrapper.classes()).contain('bike-bookmark--active')

    wrapper.setProps({ active: false })
    await flushPromises()

    expect(wrapper.classes()).not.contain('bike-bookmark--active')
  })
})
