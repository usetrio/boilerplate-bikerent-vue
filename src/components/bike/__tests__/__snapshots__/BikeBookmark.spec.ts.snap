// Vitest Snapshot v1

exports[`BikeBookmark > renders correctly 1`] = `
"<div class=\\"bike-bookmark\\" data-v-1f9ffc06=\\"\\">
  <div class=\\"icon text-base\\" data-v-6ea13f32=\\"\\" data-v-1f9ffc06=\\"\\"><em class=\\"fa-regular fa-heart\\" data-v-6ea13f32=\\"\\"></em></div>
</div>"
`;
